rosbag2_compression/src/rosbag2_compression/compression_factory_impl.hpp

```diff
@@ -20,67 +20,116 @@
+};
+
+template<typename InterfaceT>
+std::shared_ptr<pluginlib::ClassLoader<InterfaceT>>
+get_class_loader()
+{
+  const char * lookup_name = CompressionTraits<InterfaceT>::name;
+  return std::make_shared<pluginlib::ClassLoader<InterfaceT>>("rosbag2_compression", lookup_name);
 }
-}  // namespace
 
-namespace rosbag2_compression
+template<typename InterfaceT>
+std::shared_ptr<InterfaceT>
+get_interface_instance(
+  std::shared_ptr<pluginlib::ClassLoader<InterfaceT>> class_loader,
+  const std::string & compression_format)
 {
+  const auto & registered_classes = class_loader->getDeclaredClasses();
+  auto class_iter = std::find(
+    registered_classes.begin(), registered_classes.end(), compression_format);
+  if (class_iter == registered_classes.end()) {
+    ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM(
+      "Requested compression format '" << compression_format << "' does not exist");
+    return nullptr;
+  }
+
+  std::shared_ptr<InterfaceT> instance = nullptr;
+  try {
+    instance = class_loader->createSharedInstance(compression_format);
+  } catch (const std::runtime_error & ex) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
+      "Unable to load instance of compression interface: " << ex.what());
+  }
+  return instance;
+}
 
 /// Implementation of the CompressionFactory. See CompressionFactory for documentation.
 class CompressionFactoryImpl
 {
 public:
-  CompressionFactoryImpl() = default;
-  ~CompressionFactoryImpl() = default;
+  CompressionFactoryImpl()
+  {
+    try {
+      compressor_class_loader_ = get_class_loader<BaseCompressorInterface>();
+    } catch (const std::exception & e) {
+      ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Unable to create class loader instance: " << e.what());
+      throw e;
+    }
+
+    try {
+      decompressor_class_loader_ = get_class_loader<BaseDecompressorInterface>();
+    } catch (const std::exception & e) {
+      ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Unable to create class loader instance: " << e.what());
+      throw e;
+    }
+  }
+
+  virtual ~CompressionFactoryImpl() = default;
 
   /// See CompressionFactory::create_compressor for documentation.
-  std::unique_ptr<rosbag2_compression::BaseCompressorInterface>
+  std::shared_ptr<rosbag2_compression::BaseCompressorInterface>
   create_compressor(const std::string & compression_format)
   {
-    if (case_insensitive_compare(compression_format, kCompressionFormatZstd)) {
-      return std::make_unique<rosbag2_compression::ZstdCompressor>();
-    } else {
-      std::stringstream errmsg;
-      errmsg << "Compression format \"" << compression_format << "\" is not supported.";
-      ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(errmsg.str());
-      throw std::invalid_argument{errmsg.str()};
+    auto instance = get_interface_instance(compressor_class_loader_, compression_format);
+    if (instance == nullptr) {
+      ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
+        "Could not load/open plugin for compression format '" << compression_format << "'");
```

Comment:
I think it would read cleaner if `return nullptr` here instead of sharing the same exit point regardless if `instance == nullptr`.